### PR TITLE
docs: added event-notification back end doc.

### PR DESCRIPTION
**Pull request by koolhead17**
added event-notification back end doc. 
## Description

The patch adds minio event-notification documentation with backend elastic, nats, rabbitmq, redis and PostgreSQL.

### CODE REVIEW

**Review by deekoder — CHANGES_REQUESTED** [review]
- [Minio](https://www.minio.io) server supports Amazon S3 compatible bucket event notifications. <-Plural

- The default location of Minio server configuration file is ``~/.minio/config.json``. Update the Elasticsearch configuration block in ``config.json`` as follows : <-- instead of period use :
- If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
- Step 2 should not start with "In this recipe" again. 
- Step 3 : Testing *at* Elasticsearch sounds incorrect.

**Review by deekoder — CHANGES_REQUESTED** [review]
Same changes should be applied to NATS & Postgres.

**Comment by koolhead17** [comment]
@deekoder 
Step 2 should not start with "In this recipe" again.

Step 3 : Testing at Elasticsearch sounds incorrect.

Please advice, what will be appropriate & will make the changes.

**Comment by deekoder** [comment]
>Please advice, what will be appropriate & will make the changes.
Step 2 should not start with "In this recipe" again. 
Please propose an alternative for me to review. 

Step 3 : Testing at Elasticsearch sounds incorrect.
Testing on ElasticSearch will be correct.

**Review by deekoder — CHANGES_REQUESTED** [review]
- line 35 Restart the Minio server to reflect config changes
 Drop the (in all recipes).

- Below terminal running nats.go program prints event notification on console.
Sentence fragment. Please propose a more clearer sentence.

- Below PostgresSQL terminal lists... 
Propose rephrased sentence.

- line 55 : It uses [pika python client] .. for RabbitMQ.
Ambiguous on what It is. Better to say : Pika Python Client, a library for RabbitMQ will be used in this python program.
-  Redis comes with handy command line interface.. 
Redis comes with a handy command line interface .. to print all notifications on the console

- Change line 60 to :  Below redis-cli prints event notifications on the console.

- Below PostgreSQL terminal lists table ``bucketevents`` stored log of event notification. 
Rephrase.

**Comment by harshavardhana** [comment]
ping @koolhead17

**Comment by koolhead17** [comment]
will send a PR this week.

**Review by donatello — CHANGES_REQUESTED** [review]
Each document can be structured and written better. No need to repetitively say that Minio, mc and BackendX (Redis/ES/etc) needs to be installed - that is obvious. Also, calling it a recipe in a documentation section sounds wrong, and many technical details are actually missing (such as the meaning of each service configuration parameter and differences between notification backends). 

Here is a sample template:

#### [Edited after feedback from @deekoder in the following comments]
# Sending Minio Event Notifications to BackendX

Write how notifications to this backend work - and highlight differences - for e.g. in postgresql delete object removes entry from table, but in NATS/Kafka, a delete object is just sent as a message to the service.

## Pre-requisites

To configure BackendX in your Minio server/cluster, you will need information about the BackendX service:

1. BackendX Hostname
2. BackendX portname
3. Credentials if needed for BackendX

## Recipe to configure BackendX in your Minio server/cluster

1. In each Minio server in your cluster, locate the configuration file, usually `~/.minio/config.json`. The file will already contain an empty configuration entry for `backendX`.
2. The configuration entry for `backendX` has the following fields - [list and describe the fields]
3. After filling in the details, update/copy over the changes to each config file in your Minio cluster, and restart each server.

## Recipe to configure notifications for a Minio bucket

You can configure notification settings on a bucket to send events to BackendX, using APIs in our Minio Client SDKs.

The following are steps to accomplish this using the command line [`mc` tool](https://minio.io/downloads/#minio-client):
[List steps with mc]

## Recipe to test that notifications are received by BackendX

1. Create an event on the bucket with notifications configured.
2. Show that the event is received by BackendX

**Comment by deekoder** [comment]
- We list the prerequisites so that the document is self contained. I believe it sounds repetitive to us but it is designed to serve our users well without assuming how they land on our pages. 

-  The term recipe need not be overused. However recipe & cookbook are well understood metaphors for developers to indicate what type of documents these are.

**Comment by donatello** [comment]
@deekoder 

> * We list the prerequisites so that the document is self contained. I believe it sounds repetitive to us but it is designed to serve our users well without assuming how they land on our pages.

I agree with you on this - I have updated the template above with a pre-requisites section where we just highlight information about the required backend service that receives notifications.
 
> * The term recipe need not be overused. However recipe & cookbook are well understood metaphors for developers to indicate what type of documents these are.

Ah ok, I get your point. I have incorporated this into the template above.

This template seems a bit better to me.

**Comment by koolhead17** [comment]
line 35 Restart the Minio server to reflect config changes
Drop the (in all recipes).

Below terminal running nats.go program prints event notification on console.
Sentence fragment. Please propose a more clearer sentence.

Below PostgresSQL terminal lists...
Propose rephrased sentence.

* line 55 : It uses [pika python client] .. for RabbitMQ.
Ambiguous on what It is. Better to say : Pika Python Client, a library for RabbitMQ will be used in this python program.
* Redis comes with handy command line interface..
* Redis comes with a handy command line interface .. to print all notifications on the console
* Change line 60 to : Below redis-cli prints event notifications on the console.
*Below PostgreSQL terminal lists table bucketevents stored log of event notification.
Rephrase.

@deekoder I have modified and made the amendments. Regarding the re-phrasing, please help me if the modified ones are not correct.

File: docs/bucket-notifications/README.md [hunk 1]
```diff
@@ -0,0 +1,435 @@
+[![Slack](https://slack.minio.io/slack?type=svg)](https://slack.minio.io)
+
+Minio server supports Amazon S3 compatible bucket event notification for following targets [AMQP](https://www.amqp.org/about/what), [Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/getting-started.html) , [Redis](http://redis.io/documentation), [nats.io](http://nats.io/) and [PostgreSQL](https://www.postgresql.org).
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: ..for the following targets

File: docs/bucket-notifications/README.md [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+				"mandatory": false,
+				"immediate": false,
+				"durable": false,
+				"internal": false,
+				"noWait": false,
+				"autoDeleted": false
+			}
+        }
+```
+Restart Minio server to reflect config changes. Minio supports all the exchange available in [RabbitMQ](https://www.rabbitmq.com/). For this setup we are using ``fanout`` exchange.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: - Minio supports all the exchanges available in.
- For this setup we are using the ``fanout`` exchange.

File: docs/bucket-notifications/README.md [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Move this note up where you do talk about the changes in the config.json.  These local changes is without context. Move it in front of the Restart Minio server...

File: docs/bucket-notifications/README.md (hunk 1 above)
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Minio server supports Amazon S3 compatible bucket event notification for *the* following targets

File: docs/bucket-notifications/README.md [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+## Publish Minio events via RabbitMQ 
+
+Install RabbitMQ from [here](https://www.rabbitmq.com/).
+
+### Recipe steps
+
+### Step 1: Add RabbitMQ endpoint to Minio
+
+The default location of Minio server configuration file is ``~/.minio/config.json``. Update the RabbitMQ configuration block in ``config.json`` as follows:
+
```
Comment (deekoder) [inline docs/bucket-notifications/README.md:29]: Bring the minio distributed note up here.

File: docs/bucket-notifications/README.md (hunk 2 above)
Comment (deekoder) [inline docs/bucket-notifications/README.md]: - Minio supports all the *exchanges* available in..
- For this setup, we are using *the* fanout exchange.

File: docs/bucket-notifications/README.md [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+			}
+        }
+```
+Restart Minio server to reflect config changes. Minio supports all the exchange available in [RabbitMQ](https://www.rabbitmq.com/). For this setup we are using ``fanout`` exchange.
+
+If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
+
+### Step 2: Enable bucket notification using Minio client
+
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:amqp``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: We will enable bucket event notification to trigger whenever a JPEG image is uploaded or deleted ...

File: docs/bucket-notifications/README.md [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+```
+mc mb myminio/images
+mc events add  myminio/images arn:minio:sqs:us-east-1:1:amqp    --suffix .jpg
+mc events list myminio/images
+arn:minio:sqs:us-east-1:1:amqp s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+```
+
+### Step 3: Testing on RabbitMQ
+
+Below python script waits on queue exchange ``bucketevents`` and prints event notification on console. It uses [Pika Python Client](https://www.rabbitmq.com/tutorials/tutorial-three-python.html), a library for RabbitMQ will be used in this python program.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: The python program below waits on *the* (please check I am not sure of the) queue exchange and  prints event notifications on the console. We use Pika Python Client library to do this.

File: docs/bucket-notifications/README.md [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+
+channel.basic_consume(callback,
+                      queue=queue_name,
+                      no_ack=False)
+
+channel.start_consuming()
+```
+
+
+Execute this example python script to watch for RabbitMQ events on the console.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Let's stick to python program. Drop the script.

File: docs/bucket-notifications/README.md [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:amqp``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
+
+```
+mc mb myminio/images
+mc events add  myminio/images arn:minio:sqs:us-east-1:1:amqp    --suffix .jpg
+mc events list myminio/images
+arn:minio:sqs:us-east-1:1:amqp s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+```
+
+### Step 3: Testing on RabbitMQ
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Step 1 & 2 were : Add, Enable. Therefore Step 3 : Test on RabbitMQ not Testing on RabbitMQ.

File: docs/bucket-notifications/README.md [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+                        "1": {
+                                "enable": true,
+                                "url": "http://127.0.0.1:9200",
+                                "index": "bucketevents"
+                        }
+                },
+```
+Restart Minio server to reflect config changes. ``bucketevents`` is the index used by Elasticsearch.
+
+If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: This line here is hanging with no context. User will not understand what "these" local changes on all nodes are.
Move it before the Restart Minio server to reflect..

File: docs/bucket-notifications/README.md [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+
+### Step 2: Enable bucket notification using Minio client
+
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:elasticsearch``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Same changes in this line as suggested for RabbitMQ.
 - We will enable bucket event notification to trigger whenever a JPEG image is uploaded or deleted ...

File: docs/bucket-notifications/README.md [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+
+### Step 3: Testing on Elasticsearch
+
+Upload a JPEG image into ``images`` bucket, this is the bucket which has been configured for event notification.
+
+```
+mc cp myphoto.jpg myminio/images
+```
+
+Running ``curl`` we can see Elasticsearch has created a new index name ``bucketevents``.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Change Running to Run (reason : You Upload a jpeg. Same tense).

File: docs/bucket-notifications/README.md [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:elasticsearch``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
+
+```
+mc mb myminio/images
+mc events add  myminio/images arn:minio:sqs:us-east-1:1:elasticsearch --suffix .jpg
+mc events list myminio/images
+arn:minio:sqs:us-east-1:1:elasticsearch s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+```
+
+### Step 3: Testing on Elasticsearch
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Testing should be Test on Elasticsearch. (Keep the tense)

File: docs/bucket-notifications/README.md [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+				"enable": true,
+				"address": "127.0.0.1:6379",
+				"password": "yoursecret",
+				"key": "bucketevents"
+			}
+		}
+```
+Restart Minio server to reflect config changes. ``bucketevents`` is the key used by Redis in this example.
+
+If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Move up before Restart Minio server.

File: docs/bucket-notifications/README.md [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+
+### Step 2: Enable bucket notification using Minio client
+
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:redis``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Same changes as mention for Elasticsearch and RabbitMQ.

File: docs/bucket-notifications/README.md [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+
+```
+mc mb myminio/images
+mc events add  myminio/images arn:minio:sqs:us-east-1:1:redis --suffix .jpg
+mc events list myminio/images
+arn:minio:sqs:us-east-1:1:redis s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+```
+
+### Step 3: Testing on Redis
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Test

File: docs/bucket-notifications/README.md [hunk 16]
```diff
@@ -0,0 +1,435 @@
...
+
+Redis comes with handy command line interface ``redis-cli`` to print all notifications on the console.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Redis comes with *a* handy command line...

File: docs/bucket-notifications/README.md [hunk 17]
```diff
@@ -0,0 +1,435 @@
...
+                "password": "yoursecret",
+                "token": "",
+                "secure": false,
+                "pingInterval": 0
+            }
+        },
+```
+Restart Minio server to reflect config changes. ``bucketevents`` is the subject used by NATS in this example.
+
+If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Move up before Restart.

File: docs/bucket-notifications/README.md [hunk 18]
```diff
@@ -0,0 +1,435 @@
...
+
+### Step 2: Enable bucket notification using Minio client
+
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:nats``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Same changes as before

File: docs/bucket-notifications/README.md [hunk 19]
```diff
@@ -0,0 +1,435 @@
...
+
+```
+mc mb myminio/images
+mc events add  myminio/images arn:minio:sqs:us-east-1:1:nats --suffix .jpg
+mc events list myminio/images
+arn:minio:sqs:us-east-1:1:nats s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+```
+
+### Step 3: Testing on NATS
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Test

File: docs/bucket-notifications/README.md [hunk 20]
```diff
@@ -0,0 +1,435 @@
...
+                                "port": "5432",
+                                "user": "postgres",
+                                "password": "mypassword",
+                                "database": "bucketevents_db"
+                        }
+                }
+```
+Restart Minio server to reflect config changes. ``bucketevents`` is the database table used by PostgreSQL in this example.
+
+If you are running [distributed Minio](https://docs.minio.io/docs/distributed-minio-quickstart-guide), modify ``~/.minio/config.json`` with these local changes on all the nodes.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Move up before Restart.

File: docs/bucket-notifications/README.md [hunk 21]
```diff
@@ -0,0 +1,435 @@
...
+
+### Step 2: Enable bucket notification using Minio client
+
+We will enable bucket events only when JPEG images are uploaded or deleted from ``images`` bucket on ``myminio`` server. Here ARN value is ``arn:minio:sqs:us-east-1:1:postgresql``. To understand more about ARN please follow [AWS ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) documentation.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Same as before.

File: docs/bucket-notifications/README.md [hunk 22]
```diff
@@ -0,0 +1,435 @@
...
+
+```
+mc mb myminio/images
+mc events add  myminio/images arn:minio:sqs:us-east-1:1:postgresql --suffix .jpg
+mc events list myminio/images
+arn:minio:sqs:us-east-1:1:postgresql s3:ObjectCreated:*,s3:ObjectRemoved:* Filter: suffix=”.jpg”
+```
+
+### Step 3: Testing on PostgreSQL
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: Test

File: docs/bucket-notifications/README.md [hunk 23]
```diff
@@ -0,0 +1,435 @@
...
+
+Open another terminal and upload a JPEG image into ``images`` bucket.
+
+```
+mc cp myphoto.jpg myminio/images
+```
+
+Open PostgreSQL terminal to list saved event notification logs.
```
Comment (deekoder) [inline docs/bucket-notifications/README.md]: to list *the* saved event ...

**Review by deekoder — CHANGES_REQUESTED** [review]
This document is getting there. Please consider my requested changes and apply them if you agree.

File: docs/bucket-notifications/README.md (hunk 1 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 4 above)
Comment (koolhead17) [inline docs/bucket-notifications/README.md]: It is a saperate note we have added, makes sense to be kept in bottom. 

That way someone will not get confused with distributed while the example workflow explaining stand alone. 

Please advice.

File: docs/bucket-notifications/README.md (hunk 2 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 3 above)
Comment (koolhead17) [inline docs/bucket-notifications/README.md]: explained above the reason behind keeping this in last line.

File: docs/bucket-notifications/README.md (hunk 5 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 8 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 6 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 7 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 9 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: Should we not add a note in footer of the doc with mention of Distributed, instead repeating in all the event notification?

Something like:
Note: If you are running distributed Minio then all the node should have modified config.json with respective event notification configuration?

File: docs/bucket-notifications/README.md (hunk 12 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 10 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 11 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 14 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 15 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 16 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 18 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 19 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 21 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 22 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

File: docs/bucket-notifications/README.md (hunk 23 above)
Comment (koolhead17) [reply docs/bucket-notifications/README.md]: ack

**Review by deekoder — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**